packages/relay/src/lib/constants.ts

```diff
@@ -91,6 +91,9 @@ export default {
         mainnet: 0x127,
         testnet: 0x128,
         previewnet: 0x129,
-    }
+    },
+
+    // block ranges
+    MAX_BLOCK_RANGE: 5,
```

Comment:
Is this value hardcoded on the mirror node? Ideally, it should be configurable, in which case the relay should make a request at startup to determine the mirror node's config.

Reply:
No. This is a relay concept to optimize considerations of doing best effort.
Given the mirror node blocks come in every 2 seconds there's a concept that a call response for a bock even 1 block back could be wrong.
This is to say on average it's likely not the case but we'll make consideration for up to `MAX_BLOCK_RANGE` behind the latest block exposed by the mirror node.

Reply:
@Nana-EC hmm, ok, the default MAX_BLOCK_RANGE of 5 equals ~10s which may not be an issue for most cases. 

However until this `MAX_BLOCK_RANGE` is deprecated i.e. until the mirror node supports archive data requests(historical execution requests) the following issues may arise:

1. if a subgraph is deployed and attempts to index an event and part of that event handler entails an `eth_call` which specified the block of the event which exceeds the MAX_BLOCK_RANGE then this will result in `UNSUPPORTED_HISTORICAL_EXECUTION` and so the subgraph will be stuck forever or at least until the mirror node now supports historical execution requests and the relay logic has been update accordingly.
2. if an `eth_call` specifies a block that is not the `latest` but within the `MAX_BLOCK_RANGE` and the result of the `eth_call` is a function of block specific data(e.g. `block.timestamp`) then this will return an invalid result. Additionally, if state changes occurred after the specified block but before `latest`(and still within `MAX_BLOCK_RANGE`) and the `eth_call` is in this case a function of that (changed)state then this will also return invalid data.

Reply:
Yep. The 1st case is why we're trying to make a best effort and the 2nd case is why the range is small.

Due to fast block times (every 2 secs) anyone making a call to `eth_blockNumber()` that then uses the response to call `eth_call` would essentially being making a historical call if they took longer that 2 secs.
For most DApps or tools this may not be the case but there will be scenarios where you are a few blocks off. 
Which is why I added the log message to get more data.

@mshakeg what are you suggesting in place?
Currently users are always getting the latest. Is it better to get the latest even when you request a historical, or fail on historical requests that are likely wrong

Reply:
Actually whiles I was writing this I was reminded of another idea I had a while back.
Right now we're guess based off of arbitrary time with block number as a proxy.

One approach could be instead to follow up with an `api/v1/contracts/{contractIdOrAddress}/state` and use the `timestamp` field to potentially determine if there was a state change since `block.timeStamp.to`.
If no then the contracts hasn't seen any state changes and returning the current state in the mean time is okay, if so then we can fail with the  `UNSUPPORTED_HISTORICAL_EXECUTION` to highlight there's been a state change.

Will think on this and see if it's possible in a follow up PR.
For now still need to explore whether to leave existing behaviour or adopt the `UNSUPPORTED_HISTORICAL_EXECUTION` which is more disruptive.
A graph node that's indexing my not care the value is the most recent but an intended historical call flow may.

Reply:
@Nana-EC hmm, how far away is historical execution support?

---
> Is it better to get the latest even when you request a historical, or fail on historical requests that are likely wrong

For my use case, there are potentially 2 types of `eth_call`, one that depends on block specific data(such as `block.timestamp`) and for this it isn't a major issue if it uses a timestamp slightly different(within ~10s at the `MAX_BLOCK_RANGE` of 5). However, the other type of eth_call is a function on storage/state and so if state changes occurred after the block specified and before or at `latest`(but within `MAX_BLOCK_RANGE`) then this may result in incorrect data being returned corrupting the subgraph's data or even breaking it.

So to account for this type of eth_call I'd recommend NOT using `MAX_BLOCK_RANGE`, but instead determining if any state changes have occurred since the requested block using the `timestamp` in `/api/v1/contracts/{contractIdOrAddress}/state` as you suggested.

And so I think leave the existing behaviour as `UNSUPPORTED_HISTORICAL_EXECUTION` is indeed quite disruptive.

Reply:
Captured this in https://github.com/hashgraph/hedera-json-rpc-relay/issues/1299